### PR TITLE
Make to_numeric default to correct precision

### LINKED ISSUE #31364: to_numeric with errors = "coerce" is adding digits at the end
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
#problem: pandas to_numeric might give some errors when using coerce 
# it is adding digits at the end.

import pandas as pd


#minimal example
data = [{'value': '.'}, {'value': '.'}, {'value': '.'}, {'value': '.'}, {'value': '243.164'}, {'value': '245.968'}, {'value': '249.585'}, {'value': '259.745'}, {'value': '265.742'}, {'value': '272.567'}]
df = pd.DataFrame(data,columns=['value'])

df.value = pd.to_numeric(df.value,errors='coerce')

#looks as if all is good:
df.value 

#but
df.value[4]


#this can be random:
data2 = [{'value': '.'}, {'value': '.'}, {'value': '.'}, {'value': '.'}, {'value': '243.164'}, {'value': '245.968'}, {'value': '249.585'}, {'value': '259.745'}, {'value': '265.742'}, {'value': '272.567'}, {'value': '279.196'}, {'value': '280.366'}, {'value': '275.034'}, {'value': '271.351'}, {'value': '272.889'}, {'value': '270.627'}, {'value': '280.828'}, {'value': '290.383'}, {'value': '308.153'}, {'value': '319.945'}, {'value': '336.0'}, {'value': '344.09'}, {'value': '351.385'}, {'value': '356.178'}, {'value': '359.82'}, {'value': '361.03'}, {'value': '367.701'}, {'value': '380.812'}, {'value': '387.98'}, {'value': '391.749'}, {'value': '391.171'}, {'value': '385.97'}, {'value': '385.345'}, {'value': '386.121'}, {'value': '390.996'}, {'value': '399.734'}, {'value': '413.073'}, {'value': '421.532'}, {'value':
'430.221'}, {'value': '437.092'}, {'value': '439.746'}, {'value': '446.01'}, {'value': '451.191'}, {'value': '460.463'}, {'value': '469.779'}, {'value': '472.025'}, {'value': '479.49'}, {'value': '474.864'}, {'value': '467.54'}, {'value': '471.978'}]


#now 4, 36, and 47 are wrong, with different endings.
df2 = pd.DataFrame(data2,columns=['value'])

df2.value = pd.to_numeric(df2.value,errors='coerce')


df2.value[[4,36,47]].to_list()
```
#### Problem description

Current behavior:  pandas to_numeric, when using errors='coerce' seem to randomly add 
decimals at the end of the number. 

#### Expected Output

no decimals should be added at the end.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : None.None

pandas           : 0.25.3
numpy            : 1.16.4
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : 0.29.12
pytest           : 5.0.1
hypothesis       : None
sphinx           : 2.1.2
blosc            : None
feather          : None
xlsxwriter       : 1.1.8
lxml.etree       : 4.3.4
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.6.1
pandas_datareader: 0.7.4
bs4              : 4.7.1
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.3.4
matplotlib       : 3.1.2
numexpr          : 2.6.9
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
s3fs             : None
scipy            : 1.2.1
sqlalchemy       : 1.3.5
tables           : 3.5.2
xarray           : 0.12.3
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.1.8
</details>

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] closes #31364
- [x] tests added / passed
  - tools/test_to_numeric.py:test_precision
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This relates to a very old issue #8002 where the default precision for CSV files could create wrong answers in the last bit.  @jreback was involved in the PR review for that, which created the default to not be the high precision parser.

For `to_numeric()`, I switched it to use `precise_xstrtod` instead of `xstrtod`  by default.

Unknown whether there are performance implications for `to_numeric()`, although this comment https://github.com/pandas-dev/pandas/issues/17154#issuecomment-319917647 from @jorisvandenbossche indicates that maybe we should consider switching to the higher precision parser by default in `read_csv()` anyway. 

Open question as to whether performance analysis of *this* PR is needed beyond what is shown below, which seems to indicate that using `precise_xstrtod` is *faster* than `xstrtod` 

The other possibility is to use a keyword argument in `to_numeric` that would allow the higher precision parser.

### CODE REVIEW

**Comment by Dr-Irv** [comment]
So I did a performance analysis of `to_numeric` using the current way `xstrtod` and the new way `precise_xstrtod`.  The penalty is minimal.  This indicates that we can make the switch here and probably change the default in `read_csv`:
Version 1.1.1:
```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: from pandas._libs import lib

In [4]: nl=np.array(list(str(i) for i in np.random.randn(1000)),dtype='O')

In [5]: nl[:5]
Out[5]:
array(['-1.1021762980719185', '0.3910582729233139',
       '-0.24666167786443186', '-0.9074814529022244',
       '0.3332612417219568'], dtype=object)

In [6]: %timeit s=lib.maybe_convert_numeric(nl, set(), coerce_numeric=False)
210 µs ± 3.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: pd.__version__
Out[7]: '1.1.1'

In [8]: nl=np.array(list(str(i) for i in np.random.randn(100000)),dtype='O')

In [9]: %timeit s=lib.maybe_convert_numeric(nl, set(), coerce_numeric=False)
20.7 ms ± 856 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
With this PR:
```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: from pandas._libs import lib

In [4]: nl=np.array(list(str(i) for i in np.random.randn(1000)),dtype='O')

In [5]: nl[:5]
Out[5]:
array(['-0.3997469965485634', '0.5214931351322711',
       '-0.37588201420103207', '0.00999602150055729', '0.633129133079011'],
      dtype=object)

In [6]:  %timeit s=lib.maybe_convert_numeric(nl, set(), coerce_numeric=False)
221 µs ± 1.48 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: pd.__version__
Out[7]: '1.2.0.dev0+275.gea2b93c47.dirty'

In [8]: nl=np.array(list(str(i) for i in np.random.randn(100000)),dtype='O')

In [9]:  %timeit s=lib.maybe_convert_numeric(nl, set(), coerce_numeric=False)
21.5 ms ± 223 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by Dr-Irv** [comment]
Realized that comparing 1.1.1 to my patch doesn't account for compiler differences.  So compared master instead:
With master:
```python
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.2.0.dev0+272.gcf1aa9eb0.dirty'

In [3]: import numpy as np

In [4]: from pandas._libs import lib

In [5]: nl=np.array(list(str(i) for i in np.random.randn(100000)),dtype='O')

In [6]:  %timeit -r 25 -n 100 s=lib.maybe_convert_numeric(nl, set(), coerce_num
   ...: eric=False)
22.2 ms ± 276 µs per loop (mean ± std. dev. of 25 runs, 100 loops each)
```
With the patch
```python
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.2.0.dev0+275.gea2b93c47.dirty'

In [3]: from pandas._libs import lib

In [4]: import numpy as np

In [5]: nl=np.array(list(str(i) for i in np.random.randn(100000)),dtype='O')

In [6]:  %timeit -r 25 -n 100 s=lib.maybe_convert_numeric(nl, set(), coerce_num
   ...: eric=False)
21.8 ms ± 447 µs per loop (mean ± std. dev. of 25 runs, 100 loops each)
```

So this illustrates that using `precise_xstrtod` instead of `xstrtod` is *faster* !

File: pandas/tests/tools/test_to_numeric.py [hunk 1]
```diff
@@ -649,3 +649,10 @@ def test_failure_to_convert_uint64_string_to_NaN():
     ser = Series([32, 64, np.nan])
     result = to_numeric(pd.Series(["32", "64", "uint64"]), errors="coerce")
     tm.assert_series_equal(result, ser)
+
+
+def test_precision():
```
Comment (jreback) [inline pandas/tests/tools/test_to_numeric.py:654]: can you parameterize this and add the other examples from the OP (currently working an non-working ones that are fixed here).

can you rename this to test_precision_float_conversion

File: doc/source/whatsnew/v1.1.2.rst [hunk 2]
```diff
@@ -37,6 +37,7 @@ Bug fixes
 - Bug in :meth:`Series.dt.isocalendar` and :meth:`DatetimeIndex.isocalendar` that returned incorrect year for certain dates (:issue:`36032`)
 - Bug in :class:`DataFrame` indexing returning an incorrect :class:`Series` in some cases when the series has been altered and a cache not invalidated (:issue:`33675`)
 - Bug in :meth:`DataFrame.corr` causing subsequent indexing lookups to be incorrect (:issue:`35882`)
+- Bug in :func:`to_numeric` where precision was incorrect (:issue:`31364`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.2.rst:40]: float precision

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
could this type of change wait til 1.2.

File: pandas/tests/tools/test_to_numeric.py (hunk 1 above)
Comment (Dr-Irv) [reply pandas/tests/tools/test_to_numeric.py]: OK, I just copied the huge list of numbers that he had there and put it in the test.  It does fail in master, works with this patch.

**Review by jorisvandenbossche — COMMENTED** [review]
Change looks fine to me (no expert in xstrtod though), but agree with @simonjayhawkins that such a non-regression fix (which actually substantially changes the implementation), although probably safe, is better targetted for 1.2

BTW, I think you by accident added a file (maybe from running the docs? in which we should actually gitignore it), `doc/example.feather`

File: doc/source/whatsnew/v1.1.2.rst [hunk 3]
```diff
@@ -42,6 +42,7 @@ Bug fixes
 - Bug in :class:`DataFrame` indexing returning an incorrect :class:`Series` in some cases when the series has been altered and a cache not invalidated (:issue:`33675`)
 - Bug in :meth:`DataFrame.corr` causing subsequent indexing lookups to be incorrect (:issue:`35882`)
 - Bug in :meth:`import_optional_dependency` returning incorrect package names in cases where package name is different from import name (:issue:`35948`)
+- Bug in :func:`to_numeric` where float precision was incorrect (:issue:`31364`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.2.rst:45]: yeah let's move to 1.2

**Review by jreback — CHANGES_REQUESTED** [review]
pls move the whatsnew note to 1.2 & have an extra file (doc/example.feather) included if you can remove

ping on green.

**Comment by simonjayhawkins** [comment]
> Change looks fine to me (no expert in xstrtod though), but agree with @simonjayhawkins that such a non-regression fix (which actually substantially changes the implementation), although probably safe, is better targetted for 1.2

sorry for this. if we weren't so close to the release, then it would have sat in master for a while.

**Comment by jorisvandenbossche** [comment]
No need to be sorry ;) regardless of how close a 1.1.x bugfix release would be, there is no need to include this PR in a bug fix release

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Dr-Irv

**Comment by Dr-Irv** [comment]
@jreback @jorisvandenbossche Based on the tests I did, I'd like to suggest that we change the default `float_precision` argument for `read_csv` and `read_table` to be `"high"` rather than `None`.  This would address issue #17154

If you agree, I can create a PR for that

**Comment by jreback** [comment]
can u run some perf tests for csv to see how much that impacts things)

**Comment by Dr-Irv** [comment]
> can u run some perf tests for csv to see how much that impacts things)

@jreback  It shouldn't affect things, and the performance tests above show that the "high" precision is slightly faster than the current default, since that is the only thing changed for `to_numeric`.  Nevertheless, I ran the appropriate asv test:

Assuming I did the asv tests right using `asv continuous -f 1.1 upstream/master issue17154 -b ^gil.ParallelReadCSV`

```
[ 50.00%] · For pandas commit acffef2e <issue17154> (round 2/2):
[ 50.00%] ·· Building for conda-py3.8-Cython0.29.16-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
........................................
[ 50.00%] ·· Benchmarking conda-py3.8-Cython0.29.16-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 75.00%] ··· gil.ParallelReadCSV.time_read_csv                              ok
[ 75.00%] ··· ========== ============
                dtype
              ---------- ------------
                float      116±4ms
                object    15.3±0.7ms
               datetime    123±3ms
              ========== ============

[ 75.00%] · For pandas commit 8c7efd1c <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.8-Cython0.29.16-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
..
[ 75.00%] ·· Benchmarking conda-py3.8-Cython0.29.16-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· gil.ParallelReadCSV.time_read_csv                              ok
[100.00%] ··· ========== ============
                dtype
              ---------- ------------
                float      124±8ms
                object    15.6±0.3ms
               datetime    125±3ms
              ========== ============


BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

So I'll create a separate PR, OK??

**Comment by jreback** [comment]
yep sounds good

**Comment by jreback** [comment]
also can likely just deprecate the option entirely now

**Comment by Dr-Irv** [comment]
> also can likely just deprecate the option entirely now

Not sure we can deprecate because of the `round_trip` option, which corresponds to the Python string-to-double converter, so will just change the default and document it.  PR coming soon....

**Pull request merged**